editor/src/clj/editor/code_completion.clj

```diff
@@ -13,55 +13,280 @@
 ;; specific language governing permissions and limitations under the License.
 
 (ns editor.code-completion
-  (:require [clojure.set :as set]
-            [clojure.string :as string]
-            [editor.lua :as lua]
+  (:require [clj-antlr.core :as antlr]
+            [clojure.java.io :as io]
+            [clojure.spec.alpha :as s]
             [internal.util :as util]))
 
-(defn replace-alias [s alias]
-  (string/replace s #".*\." (str alias ".")))
+;; Completion spec
+(s/def ::name string?) ;; used for filtering
+(s/def ::display-string string?) ;; used for view
+(s/def :editor.code-completion.insert/type #{:snippet :plaintext})
+(s/def :editor.code-completion.insert/value string?)
+(s/def ::insert
+  (s/keys :req-un [:editor.code-completion.insert/type
+                   :editor.code-completion.insert/value]))
```

Comment:
I think this would add more confusion instead because we define `::insertion` below as post-processed data that contains information about the inserted string and tab triggers, choices, and exits. This is another entity that is used to produce the insertion, so it should have a different name.